### ISSUE TITLE
Feature request: get last modification time from git

### ISSUE THREAD

**Issue by Frank3K** [pgilad/gulp-sitemap#12]
I think it would be a nice addition if the last modification time could be taken from the git log of the corresponding html-file. When using the default option, the last modification time is taken from the file itself (if I understand correctly). For git, this is by default the time the file was created or updated using `git pull`. Furthermore, when using beautifier like [gulp-jsbeautifier](https://github.com/tarunc/gulp-jsbeautifier) the modification date is updated upon beautification.

One can get the last modification time (in ISO-8601 notation) of a file (e.g. index.html) using the following git command:

``` git
git log -1 --format=%cI index.html
```

Using plugins like [gulp-git](https://github.com/stevelacy/gulp-git) this can be embedded in a plugin as follows (pseudo-code):

``` js
  var path = ....;
  var cmd = 'log -1 --format=%cI ' + paths;
  return git.exec({ args: cmd, cwd: cwd, quiet: true }, function(err, stdout){
     if (!err && lastmod) {
      var lastmod = stdout.trim();
    }
  });
```

**Comment by pgilad** [comment]
Hey, unless you are changing the file you are working on, the mtime for change will remain constant regardless of you are using git or not. 

I think that by serving the most up-to-date mtime of the actual file is pretty exact, so I don't really understand why you want to switch it to git..

**Comment by Frank3K** [comment]
When doing a `git pull` all new files are created with a modification date of now.

Suppose the following scenario:
- Team member A adds a page `page1.html` on a given day `X`.
- Team member B does a git pull one week later. On his device, `page1.html` has a modification date of `X+1week`.
- When both A and B generate the sitemap, the times for `page1` are one week apart.

**Comment by pgilad** [comment]
I see your point now... The obvious thing would be to allow the user a way to define a function that will calculate the mtime by whatever he wants

**Comment by Frank3K** [comment]
That would be nice indeed, and it wouldn't tie this plugin to `gulp-git`. 

I'm not familiar with writing gulp plugins, so I'm not sure I can help out by creating a PR. The hardest part (I think) is to handle functions that have an asynchronously callback (like `git-gulp` does).

**Comment by Frank3K** [comment]
I think we don't need git-gulp at all when we just execute the command line git. This can be done synchronous in [node](https://nodejs.org/api/child_process.html#child_process_child_process_execsync_command_options), which makes integration easier. I'll try to come up with a PR this week.

**Comment by pgilad** [comment]
Great thanks!

**Comment by alethea** [comment]
What if you could make it a function that supplied a value to a callback, E.g.:

```javascript
{
  lastmod: function (file, cb) {
    var cmd = 'log -1 --format=%cI  --' + path;
    git.exec({ args: cmd, cwd: cwd, quiet: true }, function(err, stdout) {
      cb(!err, stdout.trim());
    }
  }
}
```

**Comment by Frank3K** [comment]
@alethea: I like your suggestion. I'm however struggling integrating this async behavior in the plugin. Do you have any hints as on how to do that properly?

**Comment by alethea** [comment]
@Frank3K: Hm, maybe make it sync, but still be a function?

```javascript
{
  lastmod: function (file) {
    var cmd = 'git log -n 1 --pretty=format:%cI  -- ' + file.relative Path;
    return new Date(execSync(cmd));
  }
}
```

**Comment by pgilad** [comment]
Thanks for the request and PR! closed via bdb86c8a07aebd5712891ceed45152ab4caae372
